[PyTorch][Static Runtime] out variant for _s_where

mikeiovine opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73438

`aten::where` has no out variant in PyTorch core, and writing one is pretty tricky because it's split into `where` and `_s_where` to make autograd stuff easier ([ref](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L4773)).

This diff implements the `where` out variant for static runtime as follows:

* Added an out variant for `_s_where` in PyTorch
* Added a static runtime implementation for `where_out` that duplicates a small amount of broadcasting logic. This function looks a lot like `where` in PyTorch, but it's slightly more efficient - static runtime can skip the device check, since we take it as given that all tensors are on CPU. We can also skip a round of dispatch by directly calling `at::cpu::_s_where_out` instead of `at::_s_where_out`.

Differential Revision: [D34469785](https://our.internmc.facebook.com/intern/diff/D34469785/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34469785/)!

Commit:
[PyTorch][Static Runtime] out variant for _s_where

`aten::where` has no out variant in PyTorch core, and writing one is pretty tricky because it's split into `where` and `_s_where` to make autograd stuff easier ([ref](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L4773)).

This diff implements the `where` out variant for static runtime as follows:

* Added an out variant for `_s_where` in PyTorch
* Added a static runtime implementation for `where_out` that duplicates a small amount of broadcasting logic. This function looks a lot like `where` in PyTorch, but it's slightly more efficient - static runtime can skip the device check, since we take it as given that all tensors are on CPU. We can also skip a round of dispatch by directly calling `at::cpu::_s_where_out` instead of `at::_s_where_out`.

Differential Revision: [D34469785](https://our.internmc.facebook.com/intern/diff/D34469785/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34469785/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Static Runtime] out variant for _s_where"


`aten::where` has no out variant in PyTorch core, and writing one is pretty tricky because it's split into `where` and `_s_where` to make autograd stuff easier ([ref](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L4773)).

This diff implements the `where` out variant for static runtime as follows:

* Added an out variant for `_s_where` in PyTorch
* Added a static runtime implementation for `where_out` that duplicates a small amount of broadcasting logic. This function looks a lot like `where` in PyTorch, but it's slightly more efficient - static runtime can skip the device check, since we take it as given that all tensors are on CPU. We can also skip a round of dispatch by directly calling `at::cpu::_s_where_out` instead of `at::_s_where_out`.

Differential Revision: [D34469785](https://our.internmc.facebook.com/intern/diff/D34469785/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34469785/)!

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -22,6 +22,18 @@ static inline void check_for_unsupported_isin_dtype(const ScalarType type) {
 
+TORCH_META_FUNC(_s_where) (const Tensor& condition, const Tensor& self, const Tensor& other) {
+  TORCH_CHECK(self.dtype() == other.dtype(), "expected scalar type ", self.dtype(), " but found ", other.dtype());
+  Tensor cond_bool = condition.scalar_type() == ScalarType::Byte ? condition.to(ScalarType::Bool) : condition;
```
This is not very good but I'm not sure there is necessarily a better way to do this. What is not good about it is that meta functions are not supposed to do any nontrivial compute, because they may get invoked as part of, e.g., just shape propagation (meta functions).  If condition happens to be a meta tensor this will be relatively cheap, but something that is supposed to work is, for example, calling meta::_s_where(cpu_tensors...) and that will useless do some compute in this case.

What makes matters a little better is that TensorIterator::build (used everywhere here) DOES do some temporary allocations, and it's kind of hacked around by just checking if *any* input argument is meta, and then avoiding allocating temporaries if it's the case. So I'm willing to let this slide, but let's have a comment+issue about it.

cc @ysiraichi

mikeiovine replied:
Sure, I'll add a comment - it's also worth noting that this byte -> bool conversion is deprecated, so the conversion will not be needed soon hopefully

ezyang reviewed (approved):
not qualified for the static runtime bit, but everything else seems to be fine. This adds meta support for `_s_where`, do we get test coverage for free?

mikeiovine commented:
> not qualified for the static runtime bit, but everything else seems to be fine. This adds meta support for `_s_where`, do we get test coverage for free?

I'm not sure, how are meta functions usually tested? Would be happy to add tests if they're missing. 

It looks like `_s_where` was removed by https://github.com/pytorch/pytorch/pull/73468, but this change was reverted, so I'm putting this PR on the back-burner for a few days until all of this stuff stabilizes.

Commit:
Update on "[PyTorch][Static Runtime] out variant for _s_where"


`aten::where` has no out variant in PyTorch core, and writing one is pretty tricky because it's split into `where` and `_s_where` to make autograd stuff easier ([ref](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L4773)).

This diff implements the `where` out variant for static runtime as follows:

* Added an out variant for `_s_where` in PyTorch
* Added a static runtime implementation for `where_out` that duplicates a small amount of broadcasting logic. This function looks a lot like `where` in PyTorch, but it's slightly more efficient - static runtime can skip the device check, since we take it as given that all tensors are on CPU. We can also skip a round of dispatch by directly calling `at::cpu::_s_where_out` instead of `at::_s_where_out`.

Differential Revision: [D34469785](https://our.internmc.facebook.com/intern/diff/D34469785/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34469785/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Static Runtime] out variant for _s_where"


`aten::where` has no out variant in PyTorch core, and writing one is pretty tricky because it's split into `where` and `_s_where` to make autograd stuff easier ([ref](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L4773)).

This diff implements the `where` out variant for static runtime as follows:

* Added an out variant for `_s_where` in PyTorch
* Added a static runtime implementation for `where_out` that duplicates a small amount of broadcasting logic. This function looks a lot like `where` in PyTorch, but it's slightly more efficient - static runtime can skip the device check, since we take it as given that all tensors are on CPU. We can also skip a round of dispatch by directly calling `at::cpu::_s_where_out` instead of `at::_s_where_out`.

Differential Revision: [D34469785](https://our.internmc.facebook.com/intern/diff/D34469785/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34469785/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Static Runtime] out variant for _s_where"


`aten::where` has no out variant in PyTorch core, and writing one is pretty tricky because it's split into `where` and `_s_where` to make autograd stuff easier ([ref](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L4773)).

This diff implements the `where` out variant for static runtime as follows:

* Added an out variant for `_s_where` in PyTorch
* Added a static runtime implementation for `where_out` that duplicates a small amount of broadcasting logic. This function looks a lot like `where` in PyTorch, but it's slightly more efficient - static runtime can skip the device check, since we take it as given that all tensors are on CPU. We can also skip a round of dispatch by directly calling `at::cpu::_s_where_out` instead of `at::_s_where_out`.

Differential Revision: [D34469785](https://our.internmc.facebook.com/intern/diff/D34469785/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34469785/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Static Runtime] out variant for _s_where"


`aten::where` has no out variant in PyTorch core, and writing one is pretty tricky because it's split into `where` and `_s_where` to make autograd stuff easier ([ref](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L4773)).

This diff implements the `where` out variant for static runtime as follows:

* Added an out variant for `_s_where` in PyTorch
* Added a static runtime implementation for `where_out` that duplicates a small amount of broadcasting logic. This function looks a lot like `where` in PyTorch, but it's slightly more efficient - static runtime can skip the device check, since we take it as given that all tensors are on CPU. We can also skip a round of dispatch by directly calling `at::cpu::_s_where_out` instead of `at::_s_where_out`.

Differential Revision: [D34469785](https://our.internmc.facebook.com/intern/diff/D34469785/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34469785/)!

[ghstack-poisoned]
